### DISCUSSION TITLE
How to pass image metadata though "colour.write_image" definition?

### DISCUSSION THREAD

**Discussion by ahemberger** [colour-science/colour-hdri#35] [Q&A]
Hi;

I'm wrestling with carrying some camera RAW images through an HDRI assembly process using ptGUI and Photoshop, whereby I'm trying to keep everything in ACES2065-1 throughout. 

Step 1 of my process is bringing my .CR2 files into ptgui, merging and stitching, and outputting a (poorly-exposed, but correct) HDR panorama. ptGUI allows me to specify an output colorspace, and so I choose ACES2065-1. This works pretty swimmingly: if I output a tiff, ACES is included as an ICC profile. If I output an EXR, ACES is denoted by the "chromaticities" metadata tag. 

My next step is to use colour and some python to apply a global exposure compensation, to bring the images into a more pleasing/useful viewable range. Here is where things go sideways on me: I seem to lose the ICC profile or the chromaticities tag when outputting my modified image:

```
oiiotool -v --stats --info original_image.exr 
Reading original_image.exr
original_image.exr : 8230 x 4115, 3 channel, float openexr
    channel list: R, G, B
    **chromaticities: 0.7347, 0.2653, 0, 1, 0.0001, -0.077, 0.32168, 0.33767**
    compression: "piz"
    oiio:ColorSpace: "Linear"
    oiio:subimages: 1
    ...
```

```
oiiotool -v --stats --info modified_image.exr 
Reading modified_image.exr
modified_image.exr : 8230 x 4115, 3 channel, float openexr
    channel list: R, G, B
    oiio:ColorSpace: "Linear"
    oiio:subimages: 1
    ...
```
Step 3 for me is to pass this image into Photoshop, to clone out tripod shadows and so on. I'm told on the ptGUI forums that the 'chromaticities' attribute is not widely supported, which encourages me to lean towards using TIFFs to navigate this process (though I would much prefer to stick with EXRs). But in either case, Photoshop doesn't seem to understand what the colorspace is of the image being loaded. There is a prompt in which I'm invited to manually specify the colorspace, but I'm a little nervous about how "correct" this is (I'm trying to avoid any sneaky/unwanted conversion of my data). So I think I would feel better about things if I could ensure the exposure-corrected file at least carries forward its ACES metadata through that step. 

I'm fuzzy on how I might successfully pass through information about my colorspace through my second step in Python. I'm using openImageIO, and my code looks like:

```
       # Read original image
        img = colour.read_image(os.path.abspath(file))
  
       # Calculate an exposure compensation value
       # ...
        img /= avg_weighted_luminance

        colour.write_image(img, out_file)
```

Thank you!

**Comment by ahemberger** [comment]
Some spelunking through the oiio docs has led me to an answer that I think is working:

```
        # Read original image
        (img, attribs) = colour.read_image(os.path.abspath(file), 
                                method="OpenImageIO", 
                                attributes=True )
                                
        # ... do stuff 

        colour.write_image(img, out_file,
                                method="OpenImageIO", 
                                attributes=attribs)
```

Sorry for the answer-my-own-question spam...

**Reply by thomasmansencal** [reply]
No worries, it is great for other people! It is worth noting that creating metadata only works with `OpenImageIO`.

**Reply by ahemberger** [reply]
Actually, on closer inspection: this seems to work with EXRs, but not with tiffs. 

A 32-bit tiff exported from ptgui is described thusly by colour after being read by openimageio:

```
        (img, attribs) = colour.read_image(os.path.abspath(file), 
                                method="OpenImageIO", 
                                attributes=True )
        print ("   Image attributes:")
        for p in attribs:
            print("    -", p.name, ":", p.value)
```

```
   Image attributes:
    - oiio:BitsPerSample : 32
    - XResolution : 300.0
    - YResolution : 300.0
    - ResolutionUnit : in
    - Make : Canon
    - Model : [Canon EOS 5D Mark III]
    - Software : PTGui Pro 12.11 (www.ptgui.com)
    - tiff:PhotometricInterpretation : 2
    - tiff:PlanarConfiguration : 1
    - planarconfig : contig
    - tiff:Compression : 5
    - compression : lzw
    - tiff:RowsPerStrip : 101
    - PixelAspectRatio : 1.0
    - ICCProfile : (0, 0, 2, 52, 108, 99, 109, 115, 4, 48, 0, 0, 109, 110, 116, 114, 82, 71, 66, 32, 88, 89, 90, 32, 7, 230, 0, 6, 0, 18, 0, 11, 0, 54, 0, 5, 97, 99, 115, 112, 65, 80, 80, 76, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 246, 214, 0, 1, 0, 0, 0, 0, 211, 45, 108, 99, 109, 115, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 11, 100, 101, 115, 99, 0, 0, 1, 8, 0, 0, 0, 48, 99, 112, 114, 116, 0, 0, 1, 56, 0, 0, 0, 76, 119, 116, 112, 116, 0, 0, 1, 132, 0, 0, 0, 20, 99, 104, 97, 100, 0, 0, 1, 152, 0, 0, 0, 44, 114, 88, 89, 90, 0, 0, 1, 196, 0, 0, 0, 20, 98, 88, 89, 90, 0, 0, 1, 216, 0, 0, 0, 20, 103, 88, 89, 90, 0, 0, 1, 236, 0, 0, 0, 20, 114, 84, 82, 67, 0, 0, 2, 0, 0, 0, 0, 16, 103, 84, 82, 67, 0, 0, 2, 0, 0, 0, 0, 16, 98, 84, 82, 67, 0, 0, 2, 0, 0, 0, 0, 16, 99, 104, 114, 109, 0, 0, 2, 16, 0, 0, 0, 36, 109, 108, 117, 99, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 12, 101, 110, 85, 83, 0, 0, 0, 20, 0, 0, 0, 28, 0, 65, 0, 67, 0, 69, 0, 83, 0, 50, 0, 48, 0, 54, 0, 53, 0, 45, 0, 49, 109, 108, 117, 99, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 12, 101, 110, 85, 83, 0, 0, 0, 48, 0, 0, 0, 28, 0, 78, 0, 111, 0, 32, 0, 99, 0, 111, 0, 112, 0, 121, 0, 114, 0, 105, 0, 103, 0, 104, 0, 116, 0, 44, 0, 32, 0, 117, 0, 115, 0, 101, 0, 32, 0, 102, 0, 114, 0, 101, 0, 101, 0, 108, 0, 121, 88, 89, 90, 32, 0, 0, 0, 0, 0, 0, 246, 214, 0, 1, 0, 0, 0, 0, 211, 45, 115, 102, 51, 50, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 88, 89, 90, 32, 0, 0, 0, 0, 0, 0, 253, 171, 0, 0, 92, 165, 255, 255, 255, 78, 88, 89, 90, 32, 0, 0, 0, 0, 255, 255, 246, 9, 255, 255, 234, 100, 0, 0, 209, 194, 88, 89, 90, 32, 0, 0, 0, 0, 0, 0, 3, 34, 0, 0, 184, 247, 0, 0, 2, 29, 112, 97, 114, 97, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 99, 104, 114, 109, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 187, 229, 0, 0, 68, 159, 0, 0, 4, 55, 0, 0, 248, 240, 255, 255, 241, 175, 255, 255, 224, 244)
    - FNumber : 22.0
    - Exif:DateTimeOriginal : 2022:05:07 09:56:19
    - Exif:DateTimeDigitized : 2022:05:07 09:56:19
    - oiio:ColorSpace : sRGB
    - exif:FNumber : 22.0
    - exif:ISOSpeedRatings : 100
    - GPano:ProjectionType : equirectangular
    - GPano:UsePanoramaViewer : 1
    - GPano:CroppedAreaImageWidthPixels : 8230
    - GPano:CroppedAreaImageHeightPixels : 4115
    - GPano:FullPanoWidthPixels : 8230
    - GPano:FullPanoHeightPixels : 4115
    - GPano:CroppedAreaLeftPixels : 0
    - GPano:CroppedAreaTopPixels : 0
    - GPano:InitialViewHeadingDegrees : 180
    - GPano:InitialViewPitchDegrees : 0
    - GPano:InitialViewRollDegrees : 0
    - GPano:StitchingSoftware : PTGui Pro 12.11 (www.ptgui.com)
    - DateTime : 2022-05-07T09:56:19
    - IPTC:MetadataDate : 2022-06-18T06:54:25-05:00
    - IPTC:ModifyDate : 2022-06-18T06:54:25-05:00
```

Then, after writing:
```
        colour.write_image(img, out_file,
                                method="OpenImageIO", 
                                attributes=attribs)
```
```

> oiiotool -v --stats --info output.tif 
 8230 x 4115, 3 channel, float tiff
    channel list: R, G, B
    compression: "lzw"
    DateTime: "2022-05-07T09:56:19"
    FNumber: 22
    Make: "Canon"
    Model: "[Canon EOS 5D Mark III]"
    Orientation: 1 (normal)
    PixelAspectRatio: 1
    planarconfig: "contig"
    ResolutionUnit: "in"
    Software: "PTGui Pro 12.11 (www.pcom)"
    XResolution: 300
    YResolution: 300
    Exif:ColorSpace: 1
    Exif:DateTimeDigitized: "2022:05:07 09:56:19"
    Exif:DateTimeOriginal: "2022:05:07 09:56:19"
    oiio:BitsPerSample: 32
    oiio:ColorSpace: "sRGB"
    tiff:Compression: 5
    tiff:PhotometricInterpretation: 2
    tiff:PlanarConfiguration: 1
    tiff:RowsPerStrip: 101
    Stats Min: 0.000000 0.000000 0.000000 (float)
    Stats Max: 144740.468750 151323.359375 169982.859375 (float)
    Stats Avg: 0.697503 0.722083 0.778969 (float)
    Stats StdDev: 239.200241 252.576218 266.587708 (float)
    Stats NanCount: 0 0 0 
    Stats InfCount: 0 0 0 
    Stats FiniteCount: 33866450 33866450 33866450 
    Constant: No
    Monochrome: No
```

I'm curious why I'm losing the ICC profile?

**Reply by thomasmansencal** [reply]
Unsure, admittedly, I never use ICC profiles anywhere so that use case never occurred.

**Reply by ahemberger** [reply]
All good; I understand using ICCs in this context isn't ideal. PTGUI is clearly built primarily for photographers rather than for VFX, and is hardcoded to restrict EXR export to half-float in a number of un-obvious cases, so I keep running into these dead ends and having to back up and find a new way around it to get a HDR all the way from RAW to final panorama entirely in ACES. 

It's a good excuse to learn a lot of new stuff though, so that part is nice. I think the path forward for me will involve writing some custom code to do equirectangular-to-cubic transformations, rather than relying on ptGUI for this step of the process.

**Reply by thomasmansencal** [reply]
I think that late version of PTGui do actually support Float32 output, I can check at work.

**Reply by ahemberger** [reply]
To be clear: the version I have will indeed export 32-bit float EXR panoramas. But, as I mention above, I need to do some cleanup to the panorama after it was been stitched: I need to clone out my tripod, remove a colorChecker, etc. Doing this in the equirectangular format is kind of a pain in Photoshop, so I have been using ptGUI's "export to cube faces" tool, which yields 6 images projected to a cube face, with straight lines that is much easier to use cloning tools on. 

The problem is that this sub-tool is hardcoded to only export half float EXRs. PTGui's support confirms this, and tells me they will add full float conversion for EXRs to their wish list for future development. They suggest exporting to TIFF for now, which _does_ export to full float. 

At each step of this process, I seem to run into some gotcha that requires switching file formats, which is confusing and - I worry - error prone. Adding to the confusion are non-obvious things like invisibly hardcoded half-float exports or loss of color profile hints, which I've run into now with imageio, opencolorio, and ptgui at various points. 

It's all ok, I'm just seeking simplicity as well as correctness for the workflow I'm building, so that my future self that tries to step through this process a few months from now can follow along without forgetting "Oh I need to export as a tiff instead as an exr at this step, but then need to convert back to exr at this other step, and make sure to check that chromaticities was correctly transferred to ICC to preserve the ACES colorspace throughout...". To help myself, I'd like to increase transparency and remove black boxes. 

So, at least for the equirectangular-to-cube-face step, while it's more work, I would rest a little easier knowing I have control over that step and that I'm not inadvertently corrupting my data.

**Reply by thomasmansencal** [reply]
Do you have a Nuke license? You could: Merge Brackets to HDRI to EXR & Generate LDR PNG for Stitching, Produce Stitching Template with LDR, Stitch EXR Brackets, Finish in Nuke.

**Reply by ahemberger** [reply]
Thank you! That would be sweet. Unfortunately this is all for a personal project, so I'm coming at this as a mortal, and Nuke isn't affordable for mortals as yet. Nuke Noncommercial is severely throttled and can't be used for this kind of work (unless one can be content with an HDRI that's 1920px wide). There's Natron, but development on it seems stall periodically, it is currently unstable on Mac M1s, and lacks the nifty pano transform tools Nuke has. 

The Unity technique is a good one for hobbyists, in that they managed to find a way through this forest with relatively inexpensive gear/software. But, the missing ingredient is ACES, and that's the nut I'm trying to crack here!

I'll get there though! I truly appreciate the generosity and patience you've offered to all my questions!

**Reply by thomasmansencal** [reply]
ACES should really be easy to add, it is merely a 3x3 matrix somewhere in the chain! :)
